tests/unittests/sources/test_cloudcix.py

```diff
@@ -0,0 +1,284 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+import json
+
+import responses
+from unittest import mock
+
+from cloudinit import distros, helpers, sources
+from cloudinit import url_helper as uh
+from cloudinit.sources import DataSourceCloudCIX as ds_mod
+from cloudinit.sources import InvalidMetaDataException
+from tests.unittests.helpers import ResponsesTestCase
+
+METADATA = {
+    "instance_id": "12_34",
+    "network": {
+        "interfaces": [
+            {
+                "mac_address": "ab:cd:ef:00:01:02",
+                "addresses": [
+                    "10.0.0.2/24",
+                    "192.168.0.2/24",
+                ],
+            },
+            {
+                "mac_address": "12:34:56:ab:cd:ef",
+                "addresses": [
+                    "10.10.10.2/24",
+                ],
+            },
+        ]
+    },
+}
+
+USERDATA = """#cloud-config
+runcmd:
+- [ echo Hello, World >> /etc/greeting ]
+"""
+
+
+class MockImds:
+    @staticmethod
+    def base_response(response):
+        return 200, response.headers, "Metadata enabled"
+
+    @staticmethod
+    def metadata_response(response):
+        return 200, response.headers, json.dumps(METADATA).encode()
+
+    @staticmethod
+    def userdata_response(response):
+        return 200, response.headers, USERDATA.encode()
+
+
+class MockEphemeralIPNetworkWithStateMsg:
+    @property
+    def state_msg(self):
+        return "Mock state"
+
+
+class TestDataSourceCloudCIX(ResponsesTestCase):
```

Comment:
We are trying to move away from CITestCase subclass in testing and children such as ResponsesTestCase where possible and into pytest-based unittests which provide more flexibility, fixtures and parametrization where possible. The testing here looks like it could be resolved using `@responses.activate` and the pytest `mocker` fixture in most cases.

[pytestify pkg](https://github.com/dannysepler/pytestify) may give you a leg-up in adapting these existing unittests the self.assert* to pytest primitives.  I can peek next week at this PR in depth to provide test change suggestions for any of the sticky situations. The following unit test tests/unittests/reporting/test_webhook_handler.py has examples of  mocker, responses.activate and a common "setup" fixture if needed as guidance.